apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerColumnCollection.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Cloud.Spanner.Data
+{
+    /// <summary>
+    /// Represents a collection of database columns associated with a <see cref="SpannerCommand" /> and their
+    /// respective mappings to columns in a DataSet.
+    /// </summary>
+    public sealed class SpannerColumnCollection : SpannerParameterCollection, IEnumerable<SpannerColumn>
+    {
+        /// <summary>
+        /// Creates a new <see cref="SpannerColumnCollection" />
+        /// </summary>
+        public SpannerColumnCollection() { }
+
+        /// <summary>
+        /// Creates a new <see cref="SpannerParameterCollection" />
+        /// </summary>
+        /// <param name="parameters">The initial collection of parameters</param>
+        public SpannerColumnCollection(IEnumerable<SpannerColumn> parameters)
+            : base(parameters) { }
+
+        /// <inheritdoc />
+        public new IEnumerator<SpannerColumn> GetEnumerator() => InnerList.Cast<SpannerColumn>().GetEnumerator();
```

Comment:
This worries me. What if someone has some general logic for dealing with `SpannerParameterCollection` and that logic creates and adds `SpannerParameter` instances to the collection (using either `Add(object)`, `AddRange(Array)`, or `Insert(int, object)`) and then someone passes a `SpannerColumnCollection` to it? It appears those operations will succeed since they are only trying to cast the the instances to `SpannerParameter`s. So it won't be until someone goes to enumerate over the collection here that things will blow up.

There are ways to handle this, but the general approach worries me a bit. Can any of these alternatives work?
- Find some name for the collection that satisfies both use cases? `SpannerFieldCollection` perhaps?
- Derive both collections from an abstract, generic, base collection.
- Have both collections contain an internal, generic, implementation collection which does the actual work and delegate to it.

I'm not going to review the rest for now in case there are other changes based on this discussion.